posthog/models/event.py

```diff
@@ -237,35 +246,58 @@ def person(self):
     # the event won't be in the Action-Event relationship yet.
     @property
     def actions(self) -> List:
+        last_updated_action_ts = Action.objects.filter(team_id=self.team_id).aggregate(models.Max("updated_at"))['updated_at__max']
+
         actions = (
             Action.objects.filter(
                 team_id=self.team_id,
-                steps__event=self.event, # filter by event name to narrow down
+                steps__event=self.event,  # filter by event name to narrow down
                 deleted=False
             )
-            .distinct("id")
-            .prefetch_related(
+                .distinct("id")
+                .prefetch_related(
                 Prefetch("steps", queryset=ActionStep.objects.order_by("id"))
             )
         )
-        if len(actions) == 0:
-            return []
-        events: models.QuerySet[Any] = Event.objects.filter(pk=self.pk)
-        for action in actions:
-            events = events.annotate(
-                **{
-                    "action_{}".format(action.pk): Event.objects.filter(pk=self.pk)
-                    .query_db_by_action(action)
-                    .values("id")[:1]
-                }
-            )
-        event = [event for event in events][0]
+        if not self.team_id in LAST_UPDATED_TEAM_ACTION \
+            or not self.team_id in TEAM_EVENT_ACTION_QUERY_CACHE \
+            or not self.event in TEAM_EVENT_ACTION_QUERY_CACHE[self.team_id] \
+            or not self.team_id in TEAM_ACTION_QUERY_CACHE \
+            or (last_updated_action_ts > LAST_UPDATED_TEAM_ACTION[self.team_id]):
+            # If actions have changed update the query
+            TEAM_ACTION_QUERY_CACHE[self.team_id], _ = actions.query.sql_with_params()
+            if len(actions) == 0:
+                return []
+            events: models.QuerySet[Any] = Event.objects.filter(pk=self.pk)
+            for action in actions:
+                events = events.annotate(
+                    **{
+                        "action_{}".format(action.pk): Event.objects.filter(pk=self.pk)
+                        .query_db_by_action(action)
+                        .values("id")[:1]
+                    }
+                )
+            q, p = events.query.sql_with_params()
+            qp = tuple(['%s' if i == self.pk else i for i in p])
+            qcache = {self.event: (q, qp)}
```

Comment:
These 2 lines are a bit magic (especially with the non-descriptive variable names). Is there a way of doing this that's slightly less magic or just more descriptive?

Reply:
Totally agree. I'll work that in too.